source/common/network/dns_impl.cc

```diff
@@ -104,16 +104,24 @@ void DnsResolverImpl::PendingResolution::onAresGetAddrInfoCallback(int status, i
     // ARES_ECONNREFUSED. If the PendingResolution has not been cancelled that means that the
     // callback_ target _should_ still be around. In that case, raise the callback_ so the target
     // can be done with this query and initiate a new one.
-    if (!cancelled_) {
-      ENVOY_LOG_EVENT(debug, "cares_dns_resolution_destroyed", "dns resolution for {} destroyed",
-                      dns_name_);
-
-      callback_(ResolutionStatus::Failure, {});
+    ENVOY_LOG_EVENT(debug, "cares_dns_resolution_destroyed", "dns resolution for {} destroyed",
+                    dns_name_);
+
+    if (!pending_response_.address_list_.empty()) {
+      ASSERT(dns_lookup_family_ == DnsLookupFamily::All && !dual_resolution_);
```

Comment:
Here and [below](https://github.com/envoyproxy/envoy/pull/18464/files#diff-e4a2ad4e57fd03d89bc0b831c86d1d365f71f9e89789cc6ccb127cecbdc95e90R179) I considered a failure in the second resolution an overall success if the first resolution was successful, and hence fired the callback with the existing addresses and a success state. Up to discussing if this should be the behavior.